### ISSUE TITLE
spdk_iscsi_conn_stop deadlock at conn.c:755

### ISSUE THREAD

**Issue by yaahallo** [spdk/spdk#744]
No known way to reproduce this failure so far outside of our internal testing framework for our product @scalecomputing

In trying to test if https://github.com/spdk/spdk/issues/725 is resolved we ran into another unrelated hang.

```
─── Stack
[3] from 0x000000000047e2cd in spdk_iscsi_conn_stop+100 at conn.c:755
arg conn = 0x7f4c08b57000
[+]
─── Expressions
─── Source 
>>> p target
$1 = (struct spdk_iscsi_tgt_node *) 0x23202c0
>>> p target->mutex
$2 = {
  __data = {
    __lock = 2,
    __count = 1024,
    __owner = 36903981,
    __nusers = 0,
    __kind = 3,
    __spins = 32,
    __elision = 0,
    __list = {
      __prev = 0x2331c34,
      __next = 0x40000000006
    }
  },
  __size = "\002\000\000\000\000\004\000\000-\034\063\002\000\000\000\000\003\000\000\000 \000\000\000\064\034\063\002\000\000\000\000\006\000\000\000\000\004\000",
  __align = 4398046511106
}
>>> bt
#0  0x00007f4c124f151d in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f4c124ece51 in _L_lock_1022 () from /lib64/libpthread.so.0
#2  0x00007f4c124ecdf2 in pthread_mutex_lock () from /lib64/libpthread.so.0
#3  0x000000000047e2cd in spdk_iscsi_conn_stop (conn=0x7f4c08b57000) at conn.c:755
#4  0x000000000047dbcb in _spdk_iscsi_conn_destruct (conn=0x7f4c08b57000) at conn.c:538
#5  0x000000000047dce2 in spdk_iscsi_conn_destruct (conn=0x7f4c08b57000) at conn.c:576
#6  0x000000000049971c in spdk_iscsi_poll_group_poll (ctx=0x22a1b40) at iscsi_subsystem.c:1181
#7  0x00000000004a713f in spdk_thread_poll (thread=0x22a0d10, max_msgs=0) at thread.c:349
#8  0x0000000000465b9c in _spdk_reactor_run (arg=0x22a0dc0) at reactor.c:320
#9  0x0000000000465eb2 in spdk_reactors_start () at reactor.c:421
#10 0x0000000000464483 in spdk_app_start (opts=0x7ffec5e72060, start_fn=0x405481 <spdk_startup>, arg1=0x0, arg2=0x0) at app.c:680
#11 0x000000000040561c in main (argc=77, argv=0x7ffec5e721e8) at iscsi_tgt.c:112
>>>
```

I'm currently trying to figure out how to identify the current owner of the lock, as soon as I've figured that  out ill leave a new comment hopefully giving enough information to isolate the issue.

**Comment by jimharris** [comment]
I don't see where we check if there are open connections to a target node if/when a target node is explicitly deleted.  I know this is part of @yaahallo's test.

When the target node gets destroyed, the mutex gets destroyed with it.  If we later try to lock on that, it could very well result in a hang.

We might be able to come up with some tests to stress this outside of your environment.  I'll need to think about this more next week.

**Comment by shuhei-matsumoto** [comment]
@yaahallo @jimharris 

What Jim says means @yaahallo's test destroys iscsi target explicitly even when there is any active connection to the iscsi target?

**Comment by jimharris** [comment]
This is just my guess. @yaahallo can confirm.

**Comment by yaahallo** [comment]
This does appear to be correct. From what I can gather when the drive is removed rudely (kicked from kernel via `echo 1 >/sys/block/sd<whatever>/device/delete`) the rpc to delete the target node can be called before all the nodes have logged out from the iscsi target.

This should be easy enough to verify without one of our more complicated test cases. The one we used to produce this error is our "throw everything and the kitchen sink at this cluster as quickly and rudely as possible" test so it can be hard to isolate exactly what happens when it catches issues.

**Comment by shuhei-matsumoto** [comment]
When I looked deleting iSCSI target node RPC, hot remove was not stable and I didn't spend enough time on this RPC. But we may be able to improve this RPC now. I'll do this RPC next. So please wait for shortly.

**Comment by pbasler-scale** [comment]
I was able to reproduce the RPC hang without Scale's software stack using iscsiadm to set up an initiator on the same machine that is running the target.

Here's the script (replace my IPs and paths with yours):
```
#!/bin/bash

rpc="python /opt/scale/include/python/iscsi_tgtd/rpc.py"

$rpc add_portal_group 1 192.168.13.15:3260
$rpc add_initiator_group 2 ANY 192.168.13.0/24
$rpc construct_aio_bdev /dev/scale/slot0data slot0 512

while true; do
    echo "constructing target node..."
    $rpc construct_target_node slot0 slot0alias slot0:0 1:2 64 -d &> /dev/null
    sleep 2

    echo "initiator login..."
    iscsiadm -m node --login &> /dev/null
    sleep 1

    initiatorpath=/dev/$(iscsiadm -m session -P 3 | grep "Attached scsi disk" | awk '{print $4}')
    echo "starting fio on $initiatorpath..."
    fio --filename=$initiatorpath --name=$initiatorpath --iodepth=64 --direct=1 --time_based \
        --runtime=172800 --readwrite=randrw --blocksize_range=512b-1m --gtod_reduce=1 \
        --warnings-fatal --thread --verify=crc32c-intel --verify_fatal=1 --ioengine=libaio \
        > fio_$outfile.out 2>&1 &
    sleep 10

    echo "deleting target node and loggint out initiator..."
    $rpc delete_target_node iqn.2018-02.scale:slot0 &> /dev/null &
    iscsiadm -m node --logout -T iqn.2018-02.scale:slot0 &> /dev/null &
    wait

    echo "checking if rpc is hung..."
    timeout 10 $rpc get_bdevs &> /dev/null
    if [[ $? -eq 124 ]]; then
        echo "RPC IS HUNG!!!!"
        exit 1
    fi
    echo "rpc OK!"
done
```

And here's the stack trace for the iscsi_tgt once the RPC is hung:
```
Thread 2 (Thread 0x7fc89ef2c700 (LWP 15744)):
#0  0x00007fc89f458183 in epoll_wait () from /lib64/libc.so.6
#1  0x00000000004d26c4 in eal_intr_handle_interrupts (pfd=5, totalfds=1) at /local/scale-product/thirdparty/spdk/spdk/dpdk/lib/librte_eal/linuxapp/eal/eal_interrupts.c:853
#2  0x00000000004d28a1 in eal_intr_thread_main (arg=0x0) at /local/scale-product/thirdparty/spdk/spdk/dpdk/lib/librte_eal/linuxapp/eal/eal_interrupts.c:937
#3  0x00000000004e6641 in rte_thread_init (arg=0x145aff0) at /local/scale-product/thirdparty/spdk/spdk/dpdk/lib/librte_eal/common/eal_common_thread.c:163
#4  0x00007fc89f72de25 in start_thread () from /lib64/libpthread.so.0
#5  0x00007fc89f457bad in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7fc8a09cf980 (LWP 15739)):
#0  0x00007fc89f734fe0 in __pause_nocancel () from /lib64/libpthread.so.0
#1  0x00007fc89f72bba5 in __pthread_mutex_lock_full () from /lib64/libpthread.so.0
#2  0x000000000047e2cd in spdk_iscsi_conn_stop (conn=0x7fc895d88000) at conn.c:755
#3  0x000000000047dbcb in _spdk_iscsi_conn_destruct (conn=0x7fc895d88000) at conn.c:538
#4  0x000000000047dce2 in spdk_iscsi_conn_destruct (conn=0x7fc895d88000) at conn.c:576
#5  0x0000000000499749 in spdk_iscsi_poll_group_poll (ctx=0x145ee40) at iscsi_subsystem.c:1181
#6  0x00000000004a716c in spdk_thread_poll (thread=0x145dd10, max_msgs=0) at thread.c:349
#7  0x0000000000465b9c in _spdk_reactor_run (arg=0x145ddc0) at reactor.c:320
#8  0x0000000000465eb2 in spdk_reactors_start () at reactor.c:421
#9  0x0000000000464483 in spdk_app_start (opts=0x7fff7e565820, start_fn=0x405481 <spdk_startup>, arg1=0x0, arg2=0x0) at app.c:680
#10 0x000000000040561c in main (argc=77, argv=0x7fff7e5659a8) at iscsi_tgt.c:112
```

So far, it's reproduced the issue 100% of the time.

I hope this helps!

**Comment by shuhei-matsumoto** [comment]
@yaahallo @pbasler-scale 

I created the patch series that starts from https://review.gerrithub.io/#/c/spdk/spdk/+/450582/ and is made of 6 patches.

Unfortunately I haven't reproduced this issue in my environment yet too but could verify this runs correctly with your script.

 will fix the issue. Will you test if it works in your usecase?

@jimharris 

Destructing iscsi target node didn't wait for the completion of scsi device destruction and I fixed that.

**Comment by shuhei-matsumoto** [comment]
@yaahallo @pbasler-scale 

I want to correct the previous comment. It looks that I could reproduce the issue and fix it by the patch series. I will test again tomorrow.

-Shuhei

**Comment by pbasler-scale** [comment]
@shuhei-matsumoto thank you for getting these patches up!

I'm having a difficult time integrating the patches because we're based off v19.01. I'm in the process of finding the commits needed to get this working, but I'm worried I'll make a mistake doing this manually.

Any assistance you could provide in merging these patches on top of v19.01 would be greatly appreciated. Thanks again!

**Comment by shuhei-matsumoto** [comment]
@pbasler-scale Of course. I will recreate the patch on spdk 19.01 this morning

**Comment by shuhei-matsumoto** [comment]
@pbasler-scale @yaahallo 

Here they are: https://review.gerrithub.io/#/c/spdk/spdk/+/450606/
Please try them in your application.

The end of the patch series is the patch by Jim to fix another issue #725, and so please ignore it if you merge it already in your application.

**Comment by shuhei-matsumoto** [comment]
@pbasler-scale according to the discussion with Jim on IRC, the patch series is not enough. Sorry for inconvenience.

**Comment by pbasler-scale** [comment]
No problem. Let me know when you have another patch that I can test with.

**Comment by yaahallo** [comment]
Just wanted to check in to see how progress on patches for this issue are going. We saw some movement on gerrit prior to the conference but haven't seen much since then and we want to make sure this doesn't fall through the cracks.

**Comment by jimharris** [comment]
Yeah - progress stopped due to the conference and now the 19.04 release.  Let's try to get back to this later next week once the 19.04 release is done.  Thanks for checking in!

**Comment by shuhei-matsumoto** [comment]
@pbasler-scale @yaahallo 

The patch series was merged to the master branch. So I created the patch series for 19.01 based on the master. Will you cherry-pick it and test the issue?

For the another issue (only cause is unknown) I have no progress.

By the way, there is a large difference between 19.01 and master, and so I hope you will adopt 19.04.

-Shuhei

**Comment by shuhei-matsumoto** [comment]
Test CI failed for all patches but the failure is not related with your trial for now. So please ignore the failure in GerritHub. Sorry for inconvenience.

**Comment by jimharris** [comment]
@yaahallo @pbasler-scale - are you able to test with 19.04?

@shuhei-matsumoto has posted patches for 19.01, but it is a very large set of patches.  If you can post results with either the 19.01 series, or with master, we can hopefully move forward on resolution.

Thanks,

-Jim

**Comment by jwatson-scale** [comment]
Hello! My name's Jeremy, I'm a colleague of @yaahallo and @pbasler-scale. We've done some testing with 19.04 and as far as we can tell the issue is fixed, but we are having other issues with 19.04 and we're not sure if they're related. We'll create more issues as we figure out what we're seeing.

**Comment by shuhei-matsumoto** [comment]
Hi @jwatson-scale 

Thank you for your test with 19.04. Could you test with 19.01 + my patch series? As Jim says, the patch series is large. You will post additional issues for 19.04. If you don't need 19.01, we may abandon the patch series for 19.01.x.

Thanks,
Shuhei

**Comment by jwatson-scale** [comment]
@shuhei-matsumoto we will not be needing 19.01, thank you for all of your help!

-Jeremy

**Comment by shuhei-matsumoto** [comment]
Thank you @jwatson-scale @yaahallo @pbasler-scale 
I abandon patches for 19.01 and close this issue. Please open additional issues as new issues.
